gh-118915: C API: Document frame locals proxies. (#127720)
python/cpython: Doc/c-api/frame.rst

   .. versionadded:: 3.11

   .. versionchanged:: 3.13
      As part of :pep:`667`, return an instance of :c:var:`PyFrameLocalsProxy_Type`.

.. c:function:: int PyFrame_GetLineNumber(PyFrameObject *frame)

   Return the line number that *frame* is currently executing.

Frame Locals Proxies
^^^^^^^^^^^^^^^^^^^^

.. versionadded:: 3.13

The :attr:`~frame.f_locals` attribute on a :ref:`frame object <frame-objects>`
is an instance of a "frame-locals proxy". The proxy object exposes a
write-through view of the underlying locals dictionary for the frame. This
ensures that the variables exposed by ``f_locals`` are always up to date with
the live local variables in the frame itself.

See :pep:`667` for more information.

.. c:var:: PyTypeObject PyFrameLocalsProxy_Type

   The type of frame :func:`locals` proxy objects.

>    .. versionadded:: 3.13

Reviewer: It seems redundant since the section was added to Python 3.13 (`.. versionadded:: 3.13` above).

Same remark for PyFrameLocalsProxy_Check().

   .. versionadded:: 3.11

   .. versionchanged:: 3.13
      As part of :pep:`667`, return an instance of :c:var:`PyFrameLocalsProxy_Type`.

.. c:function:: int PyFrame_GetLineNumber(PyFrameObject *frame)

   Return the line number that *frame* is currently executing.

Frame Locals Proxies
^^^^^^^^^^^^^^^^^^^^

.. versionadded:: 3.13

In older versions of :term:`CPython`, the exposed local variables dictionary

> (via :attr:`~frame.f_locals`, :c:func:`PyFrame_GetLocals`, or :func:`locals`)

Reviewer 2: ```suggestion
(via :attr:`~frame.f_locals`, :c:func:`PyFrame_GetLocals` or :func:`locals`)
```

Reviewer 3: We allow the Oxford comma in general, so it's fine.

Author: This is an Oxford comma. I'm not sure what the precedent for us is, I'll double check.

Edit: Bénédikt beat me to the comment :laughing:, nevermind.

wouldn't be consistent with the actual local variables, leading to all sorts of

> weird issues. In 3.13, instead of using a dictionary to expose local variables,

Reviewer 2: ```suggestion
weird issues. In Python 3.13, instead of using a dictionary to expose local variables,
```

> In older versions of :term:`CPython`, the exposed local variables dictionary
> (via :attr:`~frame.f_locals`, :c:func:`PyFrame_GetLocals`, or :func:`locals`)
> wasn't consistent with the actual local variables, which lead to all sorts of
> weird issues. In Python 3.13, instead of using a dictionary to expose locals,
> a special proxy type that retains consistency when manipulating
> scopes is used.

Reviewer 4: I would say that it would be better for the first sentence of this paragraph to say what frame-local proxies are, rather than what the type of `frame.f_locals` _used_ to be :-) In fact, I'm not totally sure we need the history lesson here at all. Maybe something like this?

```suggestion
The :attr:`~frame.f_locals` attribute on a :ref:`frame object <frame-objects>`
is an instance of a "frame-locals proxy". The proxy object exposes a
write-through view of the underlying locals dictionary for the frame.
```

Author: I like the new wording, but I do think that explaining the prior behavior is important.

Reviewer 4: In general, we try to draw a line between PEPs and docs.python.org documentation:
- PEPs are historical documents. They record the decisions that were taken, the discussions around those decisions, and the context around those decisions. However, they are not guaranteed to be kept up to date; you should go to docs.python.org for the living documentation
- docs.python.org is living documentation that is kept up to date with Python as it currently exists. If information on docs.python.org is outdated, that's a bug that you can report on our issue tracker. However, the full context and history behind decisions that were taken is usually kept out of the reference docs, for brevity, and because it's largely irrelevant to current users. If you want that information, you should go and read the PEP.

What's the motivation here for breaking from that general principle? Why do users need to understand what came before in order to understand the behaviour now?

If you want to explain the motivations for exposing a proxy, I think that can be done without digging into the history:

```suggestion
The :attr:`~frame.f_locals` attribute on a :ref:`frame object <frame-objects>`
is an instance of a "frame-locals proxy". The proxy object exposes a
write-through view of the underlying locals dictionary for the frame. This
ensures that the variables exposed by ``f_locals`` are always up to date with
the live local variables in the frame itself.
```

Author: I'm still learning what the conventions for documentation are, so I'll just go with your suggestion. Thanks!
